route_guide/route_guide_server.cc

```diff
@@ -55,133 +66,226 @@ float GetDistance(const Point& start, const Point& end) {
+                     if (!GetFeatureName(*point, feature_list_).empty()) {
+                       feature_count++;
+                     }
+                     if (point_count != 1) {
+                       distance += GetDistance(previous, *point);
+                     }
+                     previous = *point;
+                   } else {
+                     // We will prepare the final response and send it to the
+                     // client when the client will finish all calls in order
+                     // to send all requests to the server.
+                     system_clock::time_point end_time = system_clock::now();
+                     RouteSummary summary;
+                     summary.set_point_count(point_count);
+                     summary.set_feature_count(feature_count);
+                     summary.set_distance(static_cast<long>(distance));
+                     auto secs = std::chrono::duration_cast<
+                         std::chrono::seconds>(
+                         end_time - start_time);
+                     summary.set_elapsed_time(secs.count());
+                     call.WriteAndFinish(summary, Status::OK);
+                   }
+                 });
+           }))
+        | Loop();
+  }
 
-    status = server->Serve<RouteGuide, Stream<RouteNote>, Stream<RouteNote>>(
-        "RouteChat",
-        [this](auto* call, auto&& note) {
-          if (note) {
-            std::unique_lock<std::mutex> lock(mu_);
-            for (const RouteNote& n : received_notes_) {
-              if (n.location().latitude() == note->location().latitude() &&
-                  n.location().longitude() == note->location().longitude()) {
-                call->Write(n);
-              }
-            }
-            received_notes_.push_back(*note);
-          } else {
-            call->Finish(Status::OK);
-          }
-        },
-        [](auto*, bool) {});
-
-    return status;
+  // ServeRouteChat implements bidirectional streaming RPC. The client sends
+  // the sequence of 4 messages. Each message consists of the RouteNote data
+  // structure (you can check this data structure in the
+  // protos/route_guide.proto). The server gets this stream of points and
+  // checks if the vector received_notes_ contains already this points. If
+  // so, then the server prepare the stream of this points.
+  auto ServeRouteChat(Server* server) {
+    return server->Accept<
+               RouteGuide,
+               Stream<RouteNote>,
+               Stream<RouteNote>>("RouteChat")
+        | Map(Then([this](auto&& context) {
+             return Server::Handler(std::move(context))
+                 .body([this](auto& call, auto&& note) {
+                   // Check if the request is valid.
+                   if (note) {
+                     // We need to lock mutex since the client do writes in
+                     // another thread to prevent data raises.
+                     std::unique_lock<std::mutex> lock(mu_);
```

Comment:
The client can't write to `received_notes_`, it's a private field inside the server. So the only concurrency issue we can have is from within the server.

We'd need to lock `received_notes_` if the server could be processing multiple `ServeRouteChat` RPCs at the same time. Can that happen, or can we remove the mutex?

Reply:
You are right. For now we don't need to use mutex, cause we will have the deal with one connection at a time. So we can remove the mutex.